### DISCUSSION TITLE
Static mypy type checking in my native machine

### DISCUSSION THREAD

**Discussion by gmmyung** [pyodide/pyodide#3985] [Q&A] [answered]
I would like to perform static type checking on my python code using mypy and vscode, but I'm facing a challenge because pyodide only runs in the browser. Is there a way to implement this?

**Comment by ryanking13** [comment]
Could you explain more about the problem you are facing? You can install [pyodide-py](https://pypi.org/project/pyodide-py/) from PyPI to get type hints of Pyodide objects.

**Comment by gmmyung** [comment]
Sorry for the confusion, I wasn't aware of the pyodode-py package. Adding to the question, is there a way to type check JSProxy classes generated by ```import js```?

**Reply by ryanking13** [reply]
Probably what you want is a [webtypy](https://github.com/pyodide/webtypy) project which @simonegiacomelli is working on to add support for Python type definitions for web APIs.

**Reply by simonegiacomelli** [reply, marked as answer]
@gmmyung if webtypy is the tool that could fit your use case, let me know if I can smooth out the use of it. The original intent of webtypy is to fuel the IDE typing system, to provide hints, smart completion, documentation about the Web API. But I think it could also be used with mypy: though, I never tried it myself.
I'm curious about what api you're interested :)

**Reply by gmmyung** [reply]
That is exactly what I was looking for! It seems like some browser functions are missing, but it is definitely usable. Thanks for your response.

**Reply by simonegiacomelli** [reply]
> That is exactly what I was looking for! 
Great! 

> It seems like some browser functions are missing, but it is definitely usable. Thanks for your response.

We are glad to expand the supported api surface. If you can, please share the function names that are missing. Thank you!

**Reply by hoodmane** [reply]
We should probably update pyodide to remove our custom js type stubs and use webtypy.

**Reply by gmmyung** [reply]
These are some blockers I have encountered:
- `Image()` constructor
-  `setAttribute()` returns `Unknown` instead of `null`
- Type narrowing from `HTMLElement` to `HTMLSomeElement` should exist
- `getContext(contextType)` should give us different types of drawing context depending on the `contextType` string

**Comment by tildebyte** [comment]
`Float32Array` *doesn't* resolve for me, with webtypy installed. However e.g. 'document', 'window', etc. *do*
